minhashcuda.cc

```diff
@@ -302,35 +302,32 @@ static std::vector<uint32_t> calc_best_split(
   // We evaluate each variant by the cumulative cost function.
   // Every call to mhcuda_calc() can grow the buffers a little; the cost function
   // optimizes for the number of reallocations first and the imbalance second.
-
+  if (devs.size() == 1) {
+    std::vector<uint32_t> v = {length};
+    return v;
+  }
   uint32_t ideal_split = rows[length] / devs.size();
-  std::vector<std::vector<uint32_t>> variants;
-  for (size_t devi = 0; devi < devs.size(); devi++) {
-    uint32_t row = std::upper_bound(
-        rows, rows + length + 1, ideal_split * (devi + 1)) - rows;
+  std::vector<std::vector<uint32_t>> variants = {{}};
+  uint32_t row = 0;
+  for (size_t devi = 0; devi < devs.size() - 1; devi++) {
     std::vector<std::vector<uint32_t>> fork;
-    if (row <= length) {
-      fork.assign(variants.begin(), variants.end());
-    }
-    if (!variants.empty()) {
-      for (auto &v : variants) {
-        v.push_back(row - 1);
-      }
-    } else {
-      variants.push_back({row - 1});
+    while (rows[row] < ideal_split * (devi + 1) && row < length - (devs.size() - 1 - devi)) {
```

Comment:
I expect that the bug can be fixed easier than rewriting the large code block. It is also harder to understand.